src/pilot/templates/first-boot.yaml

```diff
@@ -14,13 +14,109 @@
 
+parameters:
+
+  ComputeTunedAdmProfile:
+    description: >
+      Apply tuned profile on compute nodes.
+      Example: "tuned-adm profile virtual-host"
+    type: string
+    default: "virtual-host"
+
+  HAProxyMaxConnections:
+    description: >
+      Configures HAProxy Max Connections
+    type: number
+
+  BufferPoolSize:
+    description: >
+      Configures MariaDB BufferPoolSize
+    type: string
+    default: "dynamic"
+
+  BufferPoolInstances:
+    description: >
+      Configures MariaDB BufferPoolInstances
+    type: number
+
 resources:
   userdata:
     type: OS::Heat::MultipartMime
     properties:
       parts:
+      - config: {get_resource: set_os_limits}
+      - config: {get_resource: performance_optimization}
       - config: {get_resource: wipe_disk}
 
+  set_os_limits:
+    type: OS::Heat::CloudConfig
+    properties:
+      cloud_config:
+        bootcmd:
+        - sed -ie 's/[[:space:]]\+/ /g' /etc/security/limits.conf; grep -v '^#' /etc/security/limits.conf | grep -q '^* soft nofile' && sed -i 's/^* soft nofile.*/* soft nofile 64000/' /etc/security/limits.conf || echo '* soft nofile 64000' >> /etc/security/limits.conf
+        - sed -ie 's/[[:space:]]\+/ /g' /etc/security/limits.conf; grep -v '^#' /etc/security/limits.conf | grep -q '^* hard nofile' && sed -i 's/^* hard nofile.*/* hard nofile 64000/' /etc/security/limits.conf || echo '* hard nofile 64000' >> /etc/security/limits.conf
+        - sed -ie 's/[[:space:]]\+/ /g' /etc/security/limits.d/20-nproc.conf; grep -v '^#' /etc/security/limits.d/20-nproc.conf | grep -q '^* soft nproc' && sed -i 's/^* soft nproc.*/* soft nproc 10240/' /etc/security/limits.d/20-nproc.conf || echo '* soft nproc 10240' >> /etc/security/limits.d/20-nproc.conf
+        - sed -ie 's/[[:space:]]\+/ /g' /etc/security/limits.d/20-nproc.conf; grep -v '^#' /etc/security/limits.d/20-nproc.conf | grep -q '^root soft nproc' && sed -i 's/^root soft nproc.*/root soft nproc unlimited/' /etc/security/limits.d/20-nproc.conf || echo 'root soft nproc unlimited' >> /etc/security/limits.d/20-nproc.conf
+
+  performance_optimization:
+    type: OS::Heat::SoftwareConfig
+    properties:
+      config:
+        str_replace:
+          template: |
+            #!/bin/bash
+            if [[ $(hostname) == *compute* ]] ; then
+              echo "Performing tuned-profile"
+              tuned-adm profile _TunedProfile_
+              echo "Applied _TunedProfile_ profile"
+            fi >> /tmp/tuned-applied.log 2>&1
+
+            if [[ $(hostname) == *controller* ]] ; then
+
+              # HAProxy Optimization
+              sed -i 's/\$haproxy_default_maxconn        =.*/\$haproxy_default_maxconn        = _HAProxyMaxconn_,/' /etc/puppet/modules/tripleo/manifests/haproxy.pp
+
+              # Openstack Service - Nova Optimization
+              sed -i 's/osapi_max_limit                      \=.*/osapi_max_limit                      = 10000,/' /etc/puppet/modules/nova/manifests/api.pp
```

Comment:
I would prefer that we use ControllerExtraConfig to set these parameters when it is possible to use that method.  I know there are some parameters here where we can't do that, so we must use this method for those.